pkg/gds/admin/v2/client.go

```diff
@@ -32,17 +44,93 @@ func New(endpoint string) (_ DirectoryAdministrationClient, err error) {
+// on disk; if it finds none then it uses either a token manager or redirects the client
+// to the browser to authenticate with the server. If the access token is expired, it
+// uses the refresh token to reauthenticate and saves the resulting tokens back to disk.
+func (s *APIv2) Login(ctx context.Context) (err error) {
+	var creds *AuthReply
+	if creds, err = s.Credentials(); err != nil {
+		// No credentials were found begin login process
+		// TODO: refactor this section to allow either local token generation or online login
+		return s.GenerateCredentials()
+	}
+	s.accessToken = creds.AccessToken
+
+	// If we've successfully loaded the credentials check the access token to make sure it's still valid
+	parser := &jwt.Parser{}
+	claims := &tokens.Claims{}
+	if _, _, err = parser.ParseUnverified(creds.AccessToken, claims); err != nil {
+		s.DeleteCredentials()
+		return errors.New("access tokens unparseable, cached credentials have been deleted, please try again")
+	}
+
+	if err = claims.Valid(); err == nil {
+		return nil
+	}
+
+	// Access token is invalid attempt to reauthenticate
+	if creds, err = s.Reauthenticate(ctx, &AuthRequest{Credential: creds.RefreshToken}); err != nil {
+		s.DeleteCredentials()
+		return fmt.Errorf("could not reauthenticate (%s), cached credentials have been deleted, please try again", err)
+	}
+
+	// Save refreshed creds to disk
+	s.accessToken = creds.AccessToken
+
+	var data []byte
+	if data, err = json.Marshal(creds); err != nil {
+		return err
+	}
+
+	folders := cfgd.QueryFolders(configdir.Global)
+	folders[0].WriteFile(credentials, data)
+
+	return nil
+}
+
+//===========================================================================
+// Client Methods
+//===========================================================================
+
+// ProtectAuthenticate sets cookies for CSRF protection (not necessary in the API but good for testing)
+func (s *APIv2) ProtectAuthenticate(ctx context.Context) (err error) {
+	//  Make the HTTP request
+	var req *http.Request
+	if req, err = s.NewRequest(ctx, http.MethodGet, "/v2/authenticate", nil, nil); err != nil {
+		return err
+	}
+
+	// Execute the request and get a response
+	if _, err = s.Do(req, nil, true); err != nil {
+		return err
+	}
+
+	// NOTE: this will only work over HTTPS, not for local debugging
+	cookies := s.client.Jar.Cookies(s.endpoint)
+	for _, cookie := range cookies {
+		if cookie.Name == CSRFCookie {
+			s.csrfToken = cookie.Value
+		}
+	}
+	return nil
+}
+
 // Authenticate the the client to the Server using the supplied credentials.
-// TODO: this method should prepare the APIv2 client to send and recv authenticated requests
-// TODO: does the client need to call the ProtectAuthenticate endpoint before auth?
-func (s APIv2) Authenticate(ctx context.Context, in *AuthRequest) (out *AuthReply, err error) {
+func (s *APIv2) Authenticate(ctx context.Context, in *AuthRequest) (out *AuthReply, err error) {
```

Comment:
Note, a * method is necessary now since the `accessToken` and `csrfToken` properties are being set by the methods.